components/camel-dynamic-router/src/main/java/org/apache/camel/component/dynamicrouter/DynamicRouterEndpoint.java

```diff
@@ -140,30 +152,90 @@ protected void doInit() throws Exception {
         super.doInit();
         DynamicRouterComponent component = getDynamicRouterComponent();
         if (CONTROL_CHANNEL_NAME.equals(configuration.getChannel())) {
-            final DynamicRouterControlChannelProcessor processor = controlChannelProcessorFactorySupplier.get()
-                    .getInstance(component);
-            processor.setConfiguration(configuration);
             try {
                 // There can be multiple control actions, but we do not want to
                 // create another consumer on the control channel, so check to
                 // see if the consumer has already been created, and skip the
                 // creation of another consumer if one already exists
                 if (component.getControlChannelProcessor() == null) {
+                    DynamicRouterControlChannelProcessor processor = controlChannelProcessorFactorySupplier.get()
+                            .getInstance(component);
+                    processor.setConfiguration(configuration);
                     component.setControlChannelProcessor(processor);
                 }
             } catch (Exception e) {
                 throw new IllegalStateException("Could not create Dynamic Router endpoint", e);
             }
         } else {
-            final DynamicRouterProcessor processor = processorFactorySupplier.get()
-                    .getInstance("dynamicRouterProcessor-" + configuration.getChannel(), getCamelContext(),
-                            configuration.getRecipientMode(), configuration.isWarnDroppedMessage(),
-                            filterProcessorFactorySupplier);
-            ServiceHelper.startService(processor);
+            CamelContext camelContext = getCamelContext();
+            String routeId = configuration.getRouteId();
+            long timeout = configuration.getTimeout();
+            ErrorHandler errorHandler = new NoErrorHandler(null);
+            if (producerCache == null) {
+                producerCache = new DefaultProducerCache(this, camelContext, 1000);
+            }
+            ExecutorService aggregateExecutorService = camelContext.getExecutorServiceManager()
```

Comment:
This looks a bit wrong where you create 2x thread pool if timeout > 0

Reply:
Yes.  That was an odd oversight.  I cleaned it up.